kumquat/management/commands/letsencrypt.py

```diff
@@ -4,34 +4,153 @@
+		private_key = rsa.generate_private_key(
+			public_exponent = 65537,
+			key_size        = settings.LETSENCRYPT_ACCT_KEY_BITS,
+			backend         = default_backend()
+		)
+		key = jose.JWKRSA(key=private_key)
+		# Prepare ACME client connection with account private key
+		net         = client.ClientNetwork(key)
+		directory   = messages.Directory.from_json(
+			net.get(settings.LETSENCRYPT_ACME_SERVER).json()
+		)
+		client_acme = client.ClientV2(directory, net=net)
+		# Generate a new account and store account information locally
+		email = getattr(settings, 'KUMQUAT_EMAIL', None)
+		regr  = client_acme.new_account(
+			messages.NewRegistration.from_data(
+				email                   = email,
+				terms_of_service_agreed = True
+			)
+		)
+		# Store private key as json format
+		with open(key_file, 'w') as f:
+			f.write(key.json_dumps())
+		# Store regr information as json format
+		with open(reg_file, 'w') as f:
+			f.write(regr.json_dumps())
+
+	return key, regr
+
+def gen_csr(domain_names, pkey_pem=None):
+	if pkey_pem is None:
+		pkey = OpenSSL.crypto.PKey()
+		pkey.generate_key(OpenSSL.crypto.TYPE_RSA, settings.LETSENCRYPT_CERT_KEY_BITS)
+		pkey_pem = OpenSSL.crypto.dump_privatekey(OpenSSL.crypto.FILETYPE_PEM, pkey)
+
+	# Generate CSR based on the domains
+	csr_pem = crypto_util.make_csr(pkey_pem, domain_names)
+	return pkey_pem, csr_pem
+
+def challenge_body(new_order):
+	"""Extract authorization resource from within order resource."""
+	# Authorization Resource: authz.
+	# This object holds the offered challenges by the server and their status.
+	authz_list = new_order.authorizations
+
+	for authz in authz_list:
+		# Choosing challenge.
+		# authz.body.challenges is a set of ChallengeBody objects.
+		for i in authz.body.challenges:
+			# Find the supported challenge.
+			if isinstance(i.chall, challenges.HTTP01):
+				return i
+
+	raise Exception('HTTP-01 challenge was not offered by the CA server.')
+
+def split_fullchain(fullchain_pem):
+	re_pem         = b"(-+BEGIN (?:.+)-+[\\r\\n]+(?:[A-Za-z0-9+/=]{1,64}[\\r\\n]+)+-+END (?:.+)-+[\\r\\n]+)"
+	cert, ca       = re.findall(re_pem, str.encode(fullchain_pem))
+	return cert.decode("ascii"), ca.decode("ascii")
+
 def issue_cert():
+	# Use or generate new account for ACME API
+	key, regr = account()
+
+	net = client.ClientNetwork(key = key, account = regr)
+	directory = messages.Directory.from_json(net.get(settings.LETSENCRYPT_ACME_SERVER).json())
+	client_acme = client.ClientV2(directory, net=net)
+
 	letsencrypt_issued = False
 	for vhost in VHost.objects.filter(use_letsencrypt=True):
 		if vhost.letsencrypt_state() not in ['REQUEST', 'RENEW']:
 			continue
+
+		pkey_pem = None
+		if vhost.letsencrypt_state() is 'RENEW':
+			pkey_pem = vhost.key
+
+		# Generate a certificate request based on the vhost and aliases.
+		pkey_pem, csr_pem = gen_csr([str(vhost),] + list(vhost.vhostalias_set.values_list('alias', flat=True)), pkey_pem)
```

Comment:
Do we need to call `.punycode()` here?

Reply:
That is totally correct! I've updated the pull request!